src/useSubscription.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  ApolloClient,
+  ApolloError,
+  OperationVariables,
+  SubscriptionOptions,
+} from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { useEffect, useRef, useState } from 'react';
+import { useApolloClient } from './ApolloContext';
+import { Omit, objToKey } from './utils';
+
+export type OnSubscriptionData<TData> = (
+  options: OnSubscriptionDataOptions<TData>
+) => any;
+
+export interface OnSubscriptionDataOptions<TData> {
+  client: ApolloClient<any>;
+  subscriptionData: SubscriptionHookResult<TData>;
+}
+
+export interface SubscriptionHookOptions<TData, TVariables>
+  extends Omit<SubscriptionOptions<TVariables>, 'query'> {
+  onSubscriptionData?: OnSubscriptionData<TData>;
+}
+
+export interface SubscriptionHookResult<TData> {
+  data?: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export function useSubscription<TData = any, TVariables = OperationVariables>(
+  query: DocumentNode,
+  {
+    variables,
+    onSubscriptionData,
+    ...options
+  }: SubscriptionHookOptions<TData, TVariables> = {}
+): SubscriptionHookResult<TData> {
+  const client = useApolloClient();
+  const onSubscriptionDataRef = useRef<null | OnSubscriptionData<TData>>(null);
```

Comment:
I'd change it to:

``` 
const onSubscriptionDataRef = useRef<
    OnSubscriptionData<TData> | null | undefined
  >(null);
```

to allow `onSubscriptionData` be both `null` and `undefined`.

Reply:
Why is that a good thing? It only makes the use more annoying as you need to check for both separately.